Assets/BossRoom/Scripts/Client/UI/Lobby/JoinLobbyUI.cs

```diff
@@ -0,0 +1,160 @@
+using System.Collections.Generic;
+using BossRoom.Scripts.Shared.Infrastructure;
+using BossRoom.Scripts.Shared.Net.UnityServices.Lobbies;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace BossRoom.Scripts.Client.UI
+{
+    /// <summary>
+    /// Handles the list of LobbyButtons and ensures it stays synchronized with the lobby list from the service.
+    /// </summary>
+    public class JoinLobbyUI : ObserverBehaviour<LobbyServiceData>
+    {
+        private UIFactory m_UIFactory;
+        private LobbyUIManager m_LobbyUIManager;
+        private LobbyServiceData m_LobbyServiceData;
+
+        [SerializeField] private LobbyPanelUI m_LobbyPanelPrototype;
+
+        //
+        [SerializeField]
+        InputField m_JoinCodeField;
+
+        /// <summary>
+        /// Key: Lobby ID, Value Lobby UI
+        /// </summary>
+        Dictionary<string, LobbyPanelUI> m_LobbyButtons = new Dictionary<string, LobbyPanelUI>();
+        Dictionary<string, LocalLobby> m_LocalLobby = new Dictionary<string, LocalLobby>();
+
+        /// <summary>Contains some amount of information used to join an existing lobby.</summary>
+        LocalLobby.LobbyData m_LocalLobbySelected;
+
+
+        [Inject]
+        private void InjectDependencies(UIFactory uiFactory, LobbyUIManager lobbyUIManager, LobbyServiceData lobbyServiceData)
+        {
+            m_UIFactory = uiFactory;
+            m_LobbyUIManager = lobbyUIManager;
+            m_LobbyServiceData = lobbyServiceData;
+        }
+
+        private void Awake()
+        {
+            m_LobbyPanelPrototype.gameObject.SetActive(false);
+        }
+
+        public void LobbyPanelSelected(LocalLobby lobby)
+        {
+            m_LocalLobbySelected = lobby.Data;
+        }
+
+        public void OnLobbyCodeInputFieldChanged(string newCode)
+        {
+            if (!string.IsNullOrEmpty(newCode))
+                m_LocalLobbySelected = new LocalLobby.LobbyData(newCode.ToUpper());
+        }
+
+        public void OnJoinButtonPressed()
+        {
+            m_LobbyUIManager.JoinLobbyRequest(m_LocalLobbySelected);
+            m_LocalLobbySelected = default;
+        }
+
+        public void OnRefresh()
+        {
+            m_LobbyUIManager.QueryLobbiesRequest();
+        }
+
+        protected override void UpdateObserver(LobbyServiceData observed)
+        {
+            ///Check for new entries, We take CurrentLobbies as the source of truth
+            var previousKeys = new List<string>(m_LobbyButtons.Keys);
+
+            foreach (var codeLobby in observed.CurrentLobbies)
```

Comment:
This logic doesn't take into account the ordering of the returned lobbies which may be something like "Available Slots", probably easier to just delete and recreate or change the button components to accept lobby data and update their rendering accordingly.  Then you don't need to recreate anything!